### ISSUE TITLE
Incorrect alpha with AlphaBlend

### ISSUE THREAD

**Issue by JustinHughart** [craftworkgames/MonoGame.Extended#298]
I was trying to fade my text out using a `BitmapFont` and I noticed an issue.

## EXPECTED

The text should fade out completely and fade back in.

## ACTUAL

https://dl.dropboxusercontent.com/u/11911775/pics/incorrecttransparency.webm

So my test shows that it works properly with every alpha value for white EXCEPT 0. It draws 0 as 255 for some reason. If I specifically assign `Color.Transparent` to it `(255,255,255,0)`, it has the same issue. This issue is not present in other parts of my game that do not use MonoGame.Extended, such as the following webm that shows how the fade should work properly.

https://dl.dropboxusercontent.com/u/11911775/pics/guiselection.webm

## HOW TO REPRODUCE

The project used is MonoGame Desktop GL.

I use the following settings for `SpriteBatch.Begin`.

    public static BlendState BlendState = BlendState.AlphaBlend;
    public static SamplerState SamplerState = SamplerState.PointClamp;
    public static DepthStencilState DepthStencilState = DepthStencilState.Default;
    public static RasterizerState RasterizerState = RasterizerState.CullNone;

Then, I draw a string using `spritebatch.DrawString(bitmapfont, text, position, Color.White * alpha)` to the screen. Alpha is a value that smoothly goes between 0 and 1 over the course of 5 seconds a pass.

## NOTES

I talked with @craftworkgames about it a bit, and we think it might be a bug in MonoGame desktop GL, but there's this: 

https://github.com/craftworkgames/MonoGame.Extended/blob/develop/Source/MonoGame.Extended/Sprites/SpriteBatchExtensions.cs#L22 

`sprite.Color * sprite.Alpha`, Color should already be drawing with alpha to begin with, so Color * Alpha should be a glitch, right? He said it was to resolve another glitch, but I'm just putting that out there.

**Comment by JustinHughart** [comment]
Tried the fade demo in the other BlendStates.

Additive: On top of everything looking completely crazy, the popin for 0 still happened.
NonPremultiplied: Popin for 0 still happened. 
Opaque: To no surprise at all, there was no popin. Or transparency at all.

So with these tests, the issue persists in all applicable blend states.

**Comment by JustinHughart** [comment]
[NumberFont.zip](https://github.com/craftworkgames/MonoGame.Extended/files/577323/NumberFont.zip)

The font used, as requested by @craftworkgames .

**Comment by craftworkgames** [comment]
Okay so I did some testing

![image](https://cloud.githubusercontent.com/assets/3201643/20097461/462ca78a-a5fa-11e6-92e5-a747cf59dedc.png)

In this screenshot you can see 6 tests performed with `BlendState.AlphaBlend`. Then you can see the same 6 tests with `BlendState.NonPremultiplied`. 

Lastly, I've got some text rendered using a Paint.NET alpha layer to demonstrate what the text should probably look like using a completely different program outside MonoGame.

Here's the code I used to render the text for the tests. I've stripped out the irrelevant bits of code, but I'll attach the full source if anyone wants to run them again.

    _spriteBatch.Begin(blendState: BlendState.AlphaBlend, transformMatrix: Matrix.CreateScale(2), samplerState: SamplerState.PointClamp);
    _spriteBatch.DrawString(_bitmapFont, "blendState: BlendState.AlphaBlend", new Vector2(12, 2), Color.White);
    _spriteBatch.DrawString(_bitmapFont, "Color.White", new Vector2(12, 14), Color.White);
    _spriteBatch.DrawString(_bitmapFont, "Color.White * 0.5f", new Vector2(12, 26), Color.White * 0.5f);
    _spriteBatch.DrawString(_bitmapFont, "new Color(Color.White, 0.5f)", new Vector2(12, 38), new Color(Color.White, 0.5f));
    _spriteBatch.DrawString(_bitmapFont, "new Color(255, 255, 255, 128)", new Vector2(12, 50), new Color(255, 255, 255, 128));
    _spriteBatch.DrawString(_bitmapFont, "new Color(255, 255, 255, 0)", new Vector2(12, 62), new Color(255, 255, 255, 0));
    _spriteBatch.DrawString(_bitmapFont, "Color.White * 0.0f", new Vector2(12, 74), Color.White * 0.0f);
    _spriteBatch.End();

    _spriteBatch.Begin(blendState: BlendState.NonPremultiplied, transformMatrix: Matrix.CreateScale(2), samplerState: SamplerState.PointClamp);
    _spriteBatch.DrawString(_bitmapFont, "blendState: BlendState.NonPremultiplied", new Vector2(12, 102), Color.White);
    _spriteBatch.DrawString(_bitmapFont, "Color.White", new Vector2(12, 114), Color.White);
    _spriteBatch.DrawString(_bitmapFont, "Color.White * 0.5f", new Vector2(12, 126), Color.White * 0.5f);
    _spriteBatch.DrawString(_bitmapFont, "new Color(Color.White, 0.5f)", new Vector2(12, 138), new Color(Color.White, 0.5f));
    _spriteBatch.DrawString(_bitmapFont, "new Color(255, 255, 255, 128)", new Vector2(12, 150), new Color(255, 255, 255, 128));
    _spriteBatch.DrawString(_bitmapFont, "new Color(255, 255, 255, 0)", new Vector2(12, 162), new Color(255, 255, 255, 0));
    _spriteBatch.DrawString(_bitmapFont, "Color.White * 0.0f", new Vector2(12, 174), Color.White * 0.0f);
    _spriteBatch.End();

Some important things to notice:

 - When using `BlendState.AlphaBlend`:
    - Test 6 is the *only* test that produces a completely transparent result.
    - Test 2 is the *only* test that looks like the text in Paint.NET
    - Test 5 is *not* transparent as one might expect (hence the reason for this bug report)
    - Tests 3 & 4 appear to be working at first glance, but if you look a little closer, only the black pixels are partially transparent. They are clearly *not* like the Paint.NET text.

 - When using `BlendState.NonPremultiplied`:
    - Tests 3 & 4 look like the text in Paint.NET
    - Tests 5 & 6 are completely and correctly transparent as you'd expect.
    - Test 2 appears to be working, but actually it's a little *too* transparent. We appear to have applied the alpha *twice*. The text doesn't look like Paint.NET.

So in conclusion, the way you apply transparency is dependent on the `BlendState`. This behavior is consistent with several other sources on the internet.

 - [How to make a Texture2D 50% transparent? XNA](http://stackoverflow.com/questions/6632723/how-to-make-a-texture2d-50-transparent-xna)
 - [Premultiplied alpha in XNA Game Studio 4.0](https://blogs.msdn.microsoft.com/shawnhar/2010/04/08/premultiplied-alpha-in-xna-game-studio-4-0/)
 - [Premultiplied alpha](https://blogs.msdn.microsoft.com/shawnhar/2009/11/06/premultiplied-alpha/)

It doesn't appear to be a bug in MonoGame or MonoGame.Extended. Just a misunderstanding of how it all works.

[Game1.zip](https://github.com/craftworkgames/MonoGame.Extended/files/578087/Game1.zip)

**Comment by LithiumToast** [comment]
See https://blogs.msdn.microsoft.com/shawnhar/2010/04/08/premultiplied-alpha-in-xna-game-studio-4-0/

**Comment by craftworkgames** [comment]
@LithiumToast I already linked to that post above 😉 

> Hmm, I read your report @craftworkgames . Looking at it, a lot of the behavior seems funky to the point where I don't get why you'd still say it's working as intended after that, honestly. Especially since it's not a problem doing that exact same thing outside of bitmapfont....
Also strangely enough, your premultiplied 6 differs from my test of it, ho hum

@JustinHughart I'm willing to keep discussing it if you think I've made a mistake (it's happened before 😄). 

I couldn't find anything wrong with it, but that doesn't mean it's not broken.  If you can come up with a test or snippet of code that proves it's broken I'm more than happy to re-open the issue.

**Comment by JustinHughart** [comment]
![texttransparency](https://cloud.githubusercontent.com/assets/5400346/20160401/4cf657e8-a6ac-11e6-9ef9-17c09688463c.png)

I upgraded the test and found that they work identically. I also found the bug in my game that was causing it. So, with this, I proved that it works entirely the same in all draw modes. Better than nothing, right? The extended test is below.

[TextTransparency.zip](https://github.com/craftworkgames/MonoGame.Extended/files/582275/TextTransparency.zip)
